### ISSUE TITLE
macOS nightly fails due to a problem in the darktable-curve-tool code when building with exiv2 0.28.1

### ISSUE THREAD

**Issue by victoryforce** [darktable-org/darktable#15655]
The title says it all.

@kmilos This is probably for you as an expert in exiv2 code.

**Comment by kmilos** [comment]
I don't think this is an exiv2 issue per se - `src/common/exif.cc` seems to compile just fine from the same log?

No idea why it would stop between 0.28.0 and 0.28.1, and why the nigthy build failed and not the regular CI one.

It looks like there are some build options missing for that tool, like `-std=c++17` for example, or the order of includes matters now, or one is missing?

**Comment by zisoft** [comment]
I have seen that and immediately tried to build dt on my local machine - no issues here.

**Comment by victoryforce** [comment]
> I have seen that and immediately tried to build dt on my local machine - no issues here.

Did you build as in nightly, with `-DBUILD_CURVE_TOOLS=ON` ?

**Comment by zisoft** [comment]
Gotcha!

```
FAILED: share/darktable/tools/basecurve/CMakeFiles/darktable-curve-tool.dir/exif-wrapper.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DGDK_DISABLE_DEPRECATED -DGTK_DISABLE_DEPRECATED -DGTK_DISABLE_SINGLE_INCLUDES -DOS_OBJECT_USE_OBJC=0 -D_XOPEN_SOURCE=700 -I/Users/mario/src/darktable/build/share/darktable/tools/basecurve -I/Users/mario/src/darktable/tools/basecurve -I/usr/local/include -fopenmp-version=45 -D_DARWIN_C_SOURCE -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -Wthread-safety -Wno-unknown-pragmas -Wno-error=varargs -Wno-error=address-of-packed-member -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fPIE -Xclang -fopenmp -MD -MT share/darktable/tools/basecurve/CMakeFiles/darktable-curve-tool.dir/exif-wrapper.cpp.o -MF share/darktable/tools/basecurve/CMakeFiles/darktable-curve-tool.dir/exif-wrapper.cpp.o.d -o share/darktable/tools/basecurve/CMakeFiles/darktable-curve-tool.dir/exif-wrapper.cpp.o -c /Users/mario/src/darktable/tools/basecurve/exif-wrapper.cpp
In file included from /Users/mario/src/darktable/tools/basecurve/exif-wrapper.cpp:19:
In file included from /usr/local/include/exiv2/exif.hpp:17:
In file included from /usr/local/include/exiv2/metadatum.hpp:10:
In file included from /usr/local/include/exiv2/value.hpp:10:
In file included from /usr/local/include/exiv2/types.hpp:10:
/usr/local/include/exiv2/slice.hpp:29:4: error: expected expression
  [[nodiscard]] inline size_t size() const noexcept {
   ^
/usr/local/include/exiv2/slice.hpp:29:17: error: expected member name or ';' after declaration specifiers
  [[nodiscard]] inline size_t size() const noexcept {
                ^
/usr/local/include/exiv2/slice.hpp:20:48: error: member initializer 'begin_' does not name a non-static data member or base class
  inline SliceBase(size_t begin, size_t end) : begin_(begin), end_(end) {
                                               ^~~~~~~~~~~~~
/usr/local/include/exiv2/slice.hpp:83:20: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using iterator = typename storage_type<data_type>::iterator;
                   ^
/usr/local/include/exiv2/slice.hpp:84:26: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using const_iterator = typename storage_type<data_type>::const_iterator;
                         ^
/usr/local/include/exiv2/slice.hpp:85:22: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using value_type = typename storage_type<data_type>::value_type;
                     ^
/usr/local/include/exiv2/slice.hpp:113:4: error: expected expression
  [[nodiscard]] const_iterator cbegin() const noexcept {
   ^
/usr/local/include/exiv2/slice.hpp:113:17: error: a type specifier is required for all declarations
  [[nodiscard]] const_iterator cbegin() const noexcept {
                ^
/usr/local/include/exiv2/slice.hpp:113:17: error: duplicate member 'const_iterator'
/usr/local/include/exiv2/slice.hpp:84:9: note: previous declaration is here
  using const_iterator = typename storage_type<data_type>::const_iterator;
        ^
/usr/local/include/exiv2/slice.hpp:113:31: error: expected ';' at end of declaration list
  [[nodiscard]] const_iterator cbegin() const noexcept {
                              ^
                              ;
/usr/local/include/exiv2/slice.hpp:92:86: error: member initializer 'storage_' does not name a non-static data member or base class
  ConstSliceBase(data_type& data, size_t begin, size_t end) : SliceBase(begin, end), storage_(data, begin, end) {
                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/exiv2/slice.hpp:102:5: error: use of undeclared identifier 'rangeCheck'
    rangeCheck(index);
    ^
/usr/local/include/exiv2/slice.hpp:107:12: error: use of undeclared identifier 'storage_'
    return storage_.unsafeAt(begin_ + index);
           ^
/usr/local/include/exiv2/slice.hpp:107:30: error: use of undeclared identifier 'begin_'
    return storage_.unsafeAt(begin_ + index);
                             ^
/usr/local/include/exiv2/slice.hpp:163:50: error: using declaration cannot refer to a constructor
  using ConstSliceBase<storage_type, data_type>::ConstSliceBase;
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
/usr/local/include/exiv2/slice.hpp:164:20: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using iterator = typename ConstSliceBase<storage_type, data_type>::iterator;
                   ^
/usr/local/include/exiv2/slice.hpp:165:26: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using const_iterator = typename ConstSliceBase<storage_type, data_type>::const_iterator;
                         ^
/usr/local/include/exiv2/slice.hpp:166:22: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using value_type = typename ConstSliceBase<storage_type, data_type>::value_type;
                     ^
/usr/local/include/exiv2/slice.hpp:186:19: error: expected ';' at end of declaration list
  iterator begin() noexcept {
                  ^
                  ;
/usr/local/include/exiv2/slice.hpp:180:12: error: use of undeclared identifier 'base_type'
    return base_type::at(index);
           ^
/usr/local/include/exiv2/slice.hpp:255:20: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using iterator = typename container::iterator;
                   ^
/usr/local/include/exiv2/slice.hpp:256:26: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using const_iterator = typename container::const_iterator;
                         ^
/usr/local/include/exiv2/slice.hpp:261:22: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using value_type = typename std::remove_cv<typename container::value_type>::type;
                     ^
/usr/local/include/exiv2/slice.hpp:280:4: error: expected expression
  [[nodiscard]] const value_type& unsafeAt(size_t index) const {
   ^
/usr/local/include/exiv2/slice.hpp:280:17: error: expected member name or ';' after declaration specifiers
  [[nodiscard]] const value_type& unsafeAt(size_t index) const {
                ^
/usr/local/include/exiv2/slice.hpp:268:70: error: member initializer 'data_' does not name a non-static data member or base class
  ContainerStorage(container& data, size_t /* begin*/, size_t end) : data_(data) {
                                                                     ^~~~~~~~~~~
/usr/local/include/exiv2/slice.hpp:326:22: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using value_type = typename std::remove_cv<typename std::remove_pointer<storage_type>::type>::type;
                     ^
/usr/local/include/exiv2/slice.hpp:328:20: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using iterator = value_type*;
                   ^
/usr/local/include/exiv2/slice.hpp:329:26: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
  using const_iterator = const value_type*;
                         ^
/usr/local/include/exiv2/slice.hpp:349:4: error: expected expression
  [[nodiscard]] value_type& unsafeAt(size_t index) noexcept {
   ^
/usr/local/include/exiv2/slice.hpp:349:17: error: a type specifier is required for all declarations
  [[nodiscard]] value_type& unsafeAt(size_t index) noexcept {
```

**Comment by kmilos** [comment]
> I was just hoping that you, as an expert in exiv2 code, could figure it out faster than I could.

You give me too much credit! 😉 

If I had to guess, I'd still suspect the compiler options - the exiv2 codebase is gradually getting migrated to C++17 and C++20 and there was probably some more cleanup between 0.28.0 and 0.28.1. C++17 is default only from Clang 16, so I guess if AppleClang is still based on Clang 14 we'd need the explicit enabling of C++17, like we do under the `src` tree.

Re includes, I'd still change to follow the order as in exiv2.hpp before including that whole file: swap error.hpp before exif.hpp...

**Comment by zisoft** [comment]
> we'd need the explicit enabling of C++17, like we do under the src tree.

BINGO 👍

adding `set(CMAKE_CXX_STANDARD 17)` to `tools/basecurve/CMakeLists.txt` makes the compiler happy again.

**Comment by zisoft** [comment]
@kmilos will you create a PR for the fix or should I ?

And I will create another PR for adding curve tool and noise tool to the macOS default build script.

**Comment by kmilos** [comment]
I guess we also need to take care of this depending on exiv2 version for other platforms:

https://github.com/darktable-org/darktable/blob/657814d1660a17ede6a0cfee654ba9653e199203/src/CMakeLists.txt#L442-L444

I won't get around to this today unfortunately, so go ahead if you can!

**Comment by zisoft** [comment]
> could get tricky as Exiv2_VERSION might not be in scope for tools?

It is, because that CMakeLists.txt already contains `find_package(Exiv2 REQUIRED)`

I have now added this snippet which works on my Mac:

```
if(CMAKE_CXX_COMPILER_ID MATCHES "AppleClang" AND Exiv2_VERSION VERSION_GREATER_EQUAL "0.28.1")
    set(CMAKE_CXX_STANDARD 17)
endif()
```

Will create a PR now.

**Comment by kmilos** [comment]
I think we want this in general, not just for AppleClang. I'd copy src/CMakeLists.txt

**Comment by zisoft** [comment]
"Clang" does not match for Xcode, the compiler ID here is "AppleClang"
